common/src/int.rs

```diff
@@ -1,120 +1,631 @@
-        let c = *c;
-        if !(c.is_ascii_alphanumeric() || c == b'_') {
-            return None;
+    pub fn into_limbs_asc(self) -> Vec<u64> {
+        self.0.into_twos_complement_limbs_asc()
+    }
+
+    pub fn to_u32_digits(&self) -> (Ordering, Vec<u32>) {
+        let abs = self.inner().unsigned_abs_ref();
+        let digits: Vec<u32> = abs.to_power_of_2_digits_asc(32);
+        (self.sign(), digits)
+    }
+
+    pub fn sqrt(self) -> Self {
+        Self(self.0.floor_sqrt())
+    }
+
+    pub fn rounding_from(value: f64) -> Option<Self> {
+        (value.is_finite() && !value.is_nan())
+            .then(|| Self(value.rounding_into(RoundingMode::Down)))
+    }
+
+    pub fn rational_of(value: f64) -> Option<(Self, Self)> {
+        let sign = value >= 0.0;
+        let rational = malachite::Rational::try_from(value).ok()?;
+        let (numerator, denominator) = rational.into_numerator_and_denominator();
+        Some((
+            Self(Integer::from_sign_and_abs(sign, numerator)),
+            Self(denominator.into()),
+        ))
+    }
+
+    pub fn modpow(&self, exponent: &Self, modulus: &Self) -> Self {
+        assert!(
+            !exponent.is_negative(),
+            "negative exponentiation is not supported!"
+        );
+        assert!(
+            !modulus.is_zero(),
+            "attempt to calculate with zero modulus!"
+        );
+
+        let mut abs = self.inner().unsigned_abs_ref().mod_pow(
+            exponent.inner().unsigned_abs_ref(),
+            modulus.inner().unsigned_abs_ref(),
+        );
+
+        if abs == <Natural as malachite::num::basic::traits::Zero>::ZERO {
+            return Self::zero();
         }
 
-        if c == b'_' && last == b'_' {
-            return None;
+        let sign = modulus.is_positive();
+
+        if self.is_negative() && exponent.inner().odd() != !sign {
+            abs = modulus.inner().unsigned_abs_ref() - abs;
         }
 
-        last = c;
+        Self(Integer::from_sign_and_abs(sign, abs))
     }
-    if last == b'_' {
-        return None;
+
+    pub fn bits(&self) -> u64 {
+        self.inner().significant_bits()
     }
 
-    // parse
-    let number = if lit.is_empty() {
-        BigInt::zero()
-    } else {
-        let uint = BigUint::parse_bytes(lit, base)?;
-        BigInt::from_biguint(sign.unwrap_or(Sign::Plus), uint)
-    };
-    Some(number)
+    pub fn count_ones(&self) -> u64 {
+        self.inner().unsigned_abs_ref().count_ones()
+    }
+
+    pub fn sign(&self) -> Ordering {
+        self.inner().sign()
+    }
+
+    pub fn to_bytes_be(&self) -> (Ordering, Vec<u8>) {
+        (
+            self.sign(),
+            self.inner().unsigned_abs_ref().to_power_of_2_digits_desc(8),
+        )
+    }
+
+    pub fn to_bytes_le(&self) -> (Ordering, Vec<u8>) {
+        (
+            self.sign(),
+            self.inner().unsigned_abs_ref().to_power_of_2_digits_asc(8),
+        )
+    }
+
+    pub fn from_bytes_be(sign: bool, digits: &[u8]) -> Self {
```

Comment:
```suggestion
    pub fn from_bytes_be(sign: Sign, digits: &[u8]) -> Self {
```
BigInt will be a part of public API of RustPython.
I wish we avoid boolean for non-boolean-meaning values for public interface as documenting purpose.